File: src/main/java/me/udnek/fnafu/mechanic/camera/CameraMenu.java
package me.udnek.fnafu.mechanic.camera;

import me.udnek.fnafu.game.Game;
import me.udnek.fnafu.item.Items;
import me.udnek.fnafu.manager.GameManager;
import me.udnek.itemscoreu.custominventory.CustomInventory;
import me.udnek.itemscoreu.utils.LogUtils;
import net.kyori.adventure.text.Component;
import org.bukkit.entity.Player;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.event.inventory.InventoryCloseEvent;
import org.bukkit.event.inventory.InventoryOpenEvent;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public class CameraMenu extends CustomInventory {
    public CameraMenu(List<Camera> cameras, Component title){
        super(9*6, title);
        LogUtils.log(title);
        for (Camera camera : cameras) {
            this.inventory.setItem(camera.getTabletMenuPosition(), Items.cameraButton.getWithCamera(camera));
        }
    }

    @Override
    public void onPlayerClicksItem(InventoryClickEvent event) {
        Game game = GameManager.getManager().getGame((Player) event.getWhoClicked());
        if (game == null) return;
        game.getEventHandler().onPlayerClicksInCameraMenu(event);
    }

    @Override
    public void onPlayerClosesInventory(InventoryCloseEvent event) {
        Game game = GameManager.getManager().getGame((Player) event.getPlayer());
        if (game == null) return;
        game.getEventHandler().onPlayerClosesCameraMenu(event);
    }

    @Override
    public void onPlayerOpensInventory(InventoryOpenEvent event) {
        Game game = GameManager.getManager().getGame((Player) event.getPlayer());
        if (game == null) return;
        game.getEventHandler().onPlayerOpensCameraMenu(event);
    }

    @Override
    public int getInventorySize() {return 9*6;}

    @Override
    public Component getDisplayName() {
        return Component.empty();
    }
}

File: src/main/kotlin/me/udnek/fnafu/mechanic/camera/CameraMenu.kt
package me.udnek.fnafu.mechanic.camera

import me.udnek.coreu.custom.inventory.ConstructableCustomInventory
import me.udnek.coreu.rpgu.component.RPGUComponents
import me.udnek.fnafu.component.FnafUComponents
import me.udnek.fnafu.item.survivor.CameraButton
import me.udnek.fnafu.util.getCameraId
import me.udnek.fnafu.util.getCustom
import me.udnek.fnafu.util.getFnafU
import net.kyori.adventure.text.Component
import org.bukkit.Color
import org.bukkit.entity.Player
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.event.inventory.InventoryCloseEvent
import org.bukkit.inventory.ItemStack

open class CameraMenu : ConstructableCustomInventory{
    private var title: Component

    constructor(cameras: List<Camera>, title: Component, cameraTablet: ItemStack) : super() {
        this.title = title
        updateCameras(cameras, null, cameraTablet)
    }

    fun updateCameras(cameras: List<Camera>, useCamera: Camera?, cameraTablet: ItemStack) {
        val isCutTublet = cameraTablet.getCustom()?.components?.get(RPGUComponents.ACTIVE_ABILITY_ITEM)?.components?.get(FnafUComponents.CAMERA_TABLET_ABILITY)?.isCut ?: return
        for (camera in cameras) {
            if (isCutTublet && !camera.isCut) continue
            if (camera == useCamera) {
                inventory.setItem(useCamera.tabletMenuPosition, CameraButton.getWithCamera(useCamera, useCamera.number, Color.GREEN)) }
            else { inventory.setItem(camera.tabletMenuPosition, CameraButton.getWithCamera(camera, camera.number, Color.BLACK)) }
        }
    }

    override fun onPlayerClicksItem(event: InventoryClickEvent) {
        event.isCancelled = true
        event.currentItem?.let {
            val id = it.getCameraId() ?: return

            (event.whoClicked as Player).getFnafU()?.let {
                it.game.systems.camera.spectateCamera(it, id, it.player.inventory.getItem(1)!!)
            }
        }
    }

    override fun onPlayerClosesInventory(event: InventoryCloseEvent) {
        (event.player as Player).getFnafU()?.let { it.game.systems.camera.exitCamera(it) }
    }

    override fun getTitle(): Component? { return title }
    override fun getInventorySize(): Int { return 9 * 6 }
}
